### PR TITLE
embree: Update to 4.3.1

### LINKED ISSUE #78332: Port raycast module to Embree 4
### Godot version

4.1.beta (598378513b256e69e9b824c36136774c41cc763c)

### System information

All

### Issue description

The `raycast` module uses the thirdparty Embree library, currently version 3.13.5.

Upstream released Embree 4 as a new major, compatibility breaking release, and we need to adapt our usage of the library API if we want to use it.

The current latest version is 4.1.0. Release notes:
- https://github.com/embree/embree/releases/tag/v4.0.0
- https://github.com/embree/embree/releases/tag/v4.1.0

#73631 was an attempt at adding support for both majors at the buildsystem level, but due to the API changes the embree4 variant wouldn't compile at all so I abandoned it.

Doing this upgrade is not urgent, and once evaluating it we might find reasons not to (e.g. I see 4.1.0 just re-added support for Linux ARM64 which 3.13.5 already had, so there might be other such downgrades for our use case). Anyone is welcome to have a look.

### PR DESCRIPTION
**Pull request by Chubercik**
Fixes #78332.

https://github.com/embree/embree/releases/tag/v4.3.1

That was a big one, I'm gonna leave commits be for now for that extra development info, if everything seems A-OK, I'll push a rebase.

I'd like to ask for an extra careful review of config files - I did the update and resolved any encountered issues, so everything compiles just fine on my end, but I haven't riffled through Godot's codebase for how the lib is used, so maybe certain `#define`s should be uncommented.

Important info about moving from version 3 to version 4 of the embree library can be found here:

- https://github.com/embree/embree?tab=readme-ov-file#upgrading-from-embree-3-to-embree-4

### CODE REVIEW

**Comment by Chubercik** [comment]
For reference, the config files that Godot gets from running `modules/raycast/godot_update_embree.py` originate from here:

- `thirdparty/embree/kernels/config.h`: https://github.com/embree/embree/blob/daa8de0e714e18ad5e5c9841b67c1950d9c91c51/kernels/config.h.in
- `thirdparty/embree/include/embree4/rtcore_config.h`: https://github.com/embree/embree/blob/daa8de0e714e18ad5e5c9841b67c1950d9c91c51/kernels/rtcore_config.h.in

**Comment by Chubercik** [comment]
> For reference, the config files that Godot gets from running `modules/raycast/godot_update_embree.py` originate from [...]

Now that I think about it, maybe a better way of going about this would be to just pull them from upstream, move them into their correct positions in the file tree and apply changes through git diffs..

Waiting for someone to weigh their opinions on this.

**Comment by Angular-Angel** [comment]
Pulled the PR for this and compiled it, tried running https://github.com/godotengine/godot-demo-projects/tree/master/3d/occlusion_culling_mesh_lod with it. Worked perfectly so far as I can tell.

File: thirdparty/embree/patches/godot-changes-noexcept-the-sequel.patch
Comment (akien-mga) [inline thirdparty/embree/patches/godot-changes-noexcept-the-sequel.patch:1]: This should be merged in `godot-changes-noexcept.patch`, since it's part of the changes needed to disable exceptions.

File: thirdparty/README.md
Comment (akien-mga) [inline thirdparty/README.md:175]: I don't know why it's missing, but I'd add `LICENSE.txt` to the files we include from upstream.

**Review by akien-mga — APPROVED** [review]
Looks great to me!

**Comment by akien-mga** [comment]
> > For reference, the config files that Godot gets from running `modules/raycast/godot_update_embree.py` originate from [...]
> 
> Now that I think about it, maybe a better way of going about this would be to just pull them from upstream, move them into their correct positions in the file tree and apply changes through git diffs..
> 
> Waiting for someone to weigh their opinions on this.

That might be a good way to highlight what changes we actually make to the default config yeah.

File: thirdparty/embree/patches/godot-changes-noexcept-the-sequel.patch
Comment (Chubercik) [reply thirdparty/embree/patches/godot-changes-noexcept-the-sequel.patch:1]: Yeah, I know; it was a last-minute addition and I never bothered merging them back together - will do so now.

**Comment by AThousandShips** [comment]
(my bad re-request was unintended)

You've accidentally broken your branch in your rebase by merging it instead of rebasing

**Comment by Chubercik** [comment]
Yeah something cursed happened, should be ok now 😬

**Comment by Chubercik** [comment]
> That was a big one, I'm gonna leave commits be for now for that extra development info, if everything seems A-OK, I'll push a rebase.

I was waiting for a good-to-go signal, I guess since @akien-mga approved I can indeed rebase now.

**Comment by fire** [comment]
Not sure what to do about PullRequest.diff too_large @akien-mga?

```
could not find pull request diff: HTTP 406: Sorry, the diff exceeded the maximum number of lines (3000) (https://api.github.com/repos/godotengine/godot/pulls/88783)
PullRequest.diff too_large
```

**Comment by AThousandShips** [comment]
It's weird how it has a limit on lines when we're just getting the names, bug on the part of `gh` perhaps?

File: modules/raycast/godot_update_embree.py
Comment (Chubercik) [inline modules/raycast/godot_update_embree.py:170]: This is a very non-Pythonic way of writing a loop, but it works just fine and I wanted to throw it out there already.

Might fix later.

**Comment by AThousandShips** [comment]
I played around some with the gh tool and found something that should work instead of the diff method, can check out if I can get it to work:
```
gh pr view ${{ github.event.pull_request.number }} --json files --template '{{range .files}}{{tablerow .path}}{{end}}'
```
This will print all the files changed by a PR, with no limits, will make a PR and check it out, and see if I can make it work with this one as a base just to see

File: modules/raycast/godot_update_embree.py
Comment (AThousandShips) [inline modules/raycast/godot_update_embree.py:152]:
```suggestion
    os.rename(os.path.join(dest_dir, config_file), os.path.join(dest_dir, config_file[:-3]))
```

File: modules/raycast/godot_update_embree.py
Comment (AThousandShips) [inline modules/raycast/godot_update_embree.py:163]:
```suggestion
shutil.move(os.path.join(dest_dir, "kernels/rtcore_config.h"), os.path.join(dest_dir, ("include/embree4/")))
```

**Review by AThousandShips — COMMENTED** [review]
And try rebasing on top of my CI fix PR to see if everything runs

**Comment by Chubercik** [comment]
> And try rebasing on top of my CI fix PR to see if everything runs

And how would I do that? Rebasing on top of `master` is easy enough, but I haven't done something like this yet.

**Comment by AThousandShips** [comment]
Might be a bit finicky, so you can wait until my PR has been merged :)

**Comment by AThousandShips** [comment]
There, please rebase :)

**Comment by Chubercik** [comment]
Will do, I'm out right now so gimme like ~1h 😅

**Comment by AThousandShips** [comment]
You merged into it instead of rebasing it seems, please do:
```
git reset --hard 15553c7
git rebase -i master
git push -f
```

**Comment by AThousandShips** [comment]
You first need to update your master branch to match the upstream, your branch is now behind and doesn't include the fix

File: .github/workflows/static_checks.yml
Comment (Chubercik) [inline .github/workflows/static_checks.yml:35]: I have no idea where GH gets this from, both mine and upstream `master` branches are up to date...

Comment (Chubercik) [reply .github/workflows/static_checks.yml:35]: Oh wait

Comment (Chubercik) [reply .github/workflows/static_checks.yml:35]: God dammit

Comment (Chubercik) [reply .github/workflows/static_checks.yml:35]: One sec 🙄

Comment (AThousandShips) [reply .github/workflows/static_checks.yml:35]: You broke your branch by not rebasing it again

Comment (Chubercik) [reply .github/workflows/static_checks.yml]: There we go 🎊

Comment (Chubercik) [reply .github/workflows/static_checks.yml]: > You broke your branch by not rebasing it again

The problem was that I didn't sync my local `master` branch with origin and I got tunnel-visioned.

**Comment by Chubercik** [comment]
Checked out the same example project as @Angular-Angel, everything seems to be working as expected:

| LOD & OC off  | LOD & OC on   |
| ------------- | ------------- |
| image | image |
| image | image |

There appears to be some instability far off in the distance, but that might've already been present with Embree 3 and is not to blame on the update (needs confirmation):

embree_test_gif

**Comment by akien-mga** [comment]
> There appears to be some instability far off in the distance, but that might've already been present with Embree 3 and is not to blame on the update (needs confirmation):

That would be worth testing with 4.3-dev5 to confirm it's not a regression.

**Comment by Chubercik** [comment]
Tested with `4.3-dev5`, said instability is also present and therefore not a result of this update :)

**Comment by Calinou** [comment]
The instability is likely fixed by https://github.com/godotengine/godot/pull/86121.

**Comment by Chubercik** [comment]
@Calinou you're right, it's gone or at least it's unperceivable on that PR's build (a4f879c7924313b1c8d9eb16ee4d05d32d9ce516).

:)

**Review by fire — APPROVED** [review]
Careful review of the comment thread suggests that the embree upgrade works and that the instability is gone.

Pair with https://github.com/godotengine/godot/pull/86121

**Comment by akien-mga** [comment]
Thanks!

**Comment by Calinou** [comment]
This PR causes a performance regression:

Before | After
-|-
Screenshot_20240404_044549_before_embree_update | Screenshot_20240404_044701_after_embree_update

Should we report it upstream?

**Comment by Chubercik** [comment]
> This PR causes a performance regression as seen in the [demo project](https://github.com/godotengine/godot-demo-projects/tree/master/3d/occlusion_culling_mesh_lod) when run at a low resolution (`--resolution 640x480`)

Is the regression only present at a low resolution?

**Comment by Calinou** [comment]
> Is the regression only present at a low resolution?

It's more noticeable at low resolutions, because rendering at a low resolution avoids being GPU-bound.

**Comment by Chubercik** [comment]
Unless going from

```c++
void RaycastOcclusionCull::Scenario::_raycast(uint32_t p_idx, const RaycastThreadData *p_raycast_data) const {
	RTCIntersectContext ctx;
	rtcInitIntersectContext(&ctx);
	ctx.flags = RTC_INTERSECT_CONTEXT_FLAG_COHERENT;

	rtcIntersect16((const int *)&p_raycast_data->masks[p_idx * TILE_RAYS], ebr_scene[current_scene_idx], &ctx, &p_raycast_data->rays[p_idx]);
}
```

to

```c++
void RaycastOcclusionCull::Scenario::_raycast(uint32_t p_idx, const RaycastThreadData *p_raycast_data) const {
	RTCRayQueryContext context;
	rtcInitRayQueryContext(&context);
	RTCIntersectArguments args;
	rtcInitIntersectArguments(&args);
	args.flags = RTC_RAY_QUERY_FLAG_COHERENT;
	args.context = &context;
	rtcIntersect16((const int *)&p_raycast_data->masks[p_idx * TILE_RAYS], ebr_scene[current_scene_idx], &p_raycast_data->rays[p_idx], &args);
}
```

could be the cause of regression, I feel like it's gotta be upstream because there are no more Godot-side changes.

The changes above were required due to Embree 4's new API; two more source files had to go through similar updates.

**Pull request merged**
